server/src/main/java/org/elasticsearch/action/support/replication/PendingReplicationActions.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.action.support.replication;
+
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.action.support.RetryableAction;
+import org.elasticsearch.common.util.concurrent.ConcurrentCollections;
+
+import java.util.Map;
+
+public class PendingReplicationActions {
```

Comment:
I we only register an action when we're actively scheduling something for retry, we might not need this class at all, and just have a simple list of listeners that wait for replication group changes, each of which checks whether the set of nodes changed? We could potentially even store the set of nodes of replication targets in ReplicationGroup. This means that a simple hashset lookup would suffice on notificaion of ReplicattionGroup changes to determine if the requests is still to be kept scheduled retrying.

Reply:
I made some of these changes (set of node ids in ReplicationGroup, only register retries, etc). But I don't see the whole vision. I understand that with a limited subset of stored listeners we can iterate through them all and pass the `ReplicationGroup`. But it seems to me that there is a race between a retry being scheduled and the listener being registered. 

The `ConcurrentMaps` in `PendingReplicationActions` prevents a listener from being scheduled on a node that left the `ReplicationGroup`. If we just maintain a list of listeners, it is possible that the `ReplicationGroup` has changed as the listener is registered. Unless you want me to forcefully call the listener with the most recent view of the `ReplicationGroup` after the listener has been registered.